### PR TITLE
[LLAPI] Add stopwords list support

### PR DESCRIPTION
**Pull request by ashtul**
Related to RedisGraph/RedisGraph#1638

### CODE REVIEW

File: tests/cpptests/test_cpp_llapi.cpp [hunk 1]
```diff
@@ -820,3 +820,14 @@ TEST_F(LLApiTest, testScorer) {
   RediSearch_ResultsIteratorFree(it);
   RediSearch_DropIndex(index);
 }
+
+TEST_F(LLApiTest, testStopwords) {
+  const char *words[] = {"Redis", "Labs"};
+  RSIndex* index = RediSearch_CreateIndex("index", NULL);
+  RediSearch_CreateStopwordsList(index, words, 2);
```
Comment (oshadmi) [inline tests/cpptests/test_cpp_llapi.cpp:827]: Should we make sure it is working as expected (not found) if we call
`RediSearch_StopwordsList_Contains` between line 826 and 827? (list should be empty?)
Same question also after line 832.

File: tests/cpptests/test_cpp_llapi.cpp [hunk 2]
```diff
@@ -820,3 +820,14 @@ TEST_F(LLApiTest, testScorer) {
...
+  ASSERT_EQ(RediSearch_StopwordsList_Contains(index, words[0], strlen(words[0])), 1);
+  ASSERT_EQ(RediSearch_StopwordsList_Contains(index, words[1], strlen(words[1])), 1);
```
Comment (oshadmi) [inline tests/cpptests/test_cpp_llapi.cpp:829]: Suggest to add another call to `RediSearch_StopwordsList_Contains` with a missing word and expect it to return `0`, also when Stopwords list was created properly, e.g., after line 829

**Review by oshadmi — COMMENTED** [review]
Nice! 👍🏼

File: tests/cpptests/test_cpp_llapi.cpp (hunk 1 above)
Comment (ashtul) [reply tests/cpptests/test_cpp_llapi.cpp:827]: Good idea.
At that point, we have the default list, so we expect `is` to return `1` and `Redis` to return `0`.

**Review by oshadmi — DISMISSED** [review]
LGTM 👍🏼

**Review by MeirShpilraien — COMMENTED** [review]
I think this API should be part of the index option before creating the index. The suggested API is confusing and make the user assume stop words list can be changed after indexing data.

**Comment by ashtul** [comment]
> I think this API should be part of the index option before creating the index. The suggested API is confusing and make the user assume stop words list can be changed after indexing data.

Good point. But that would break the existing API.
We have two options
* keep it as is but declare that it is impossible to change the stopwords list after data insertion.
* break the current API by adding `const char **cstr, int len` to `RediSearch_CreateIndex`.

WDYT?

**Comment by oshadmi** [comment]
@MeirShpilraien When adding schema fields by `FT.ALTER` or when documents are being updated - wont the updated stopwords list be used then?

**Comment by MeirShpilraien** [comment]
@ashtul we do not have to break any API, we can add this API to the index option parameter that is given to the create function, this is exactly what this is for.
@oshadmi I do not think we can change the stop wards list of index alter, and regardless, on the LLAPI we can do not have index alter (and it does not make sense to have it because it requires reindexing so its the same as recreating the index..)

**Comment by ashtul** [comment]
@MeirShpilraien you are right.
Done!

File: src/redisearch_api.c [hunk 3]
```diff
@@ -247,6 +251,9 @@ int RediSearch_IndexAddDocument(IndexSpec* sp, Document* d, int options, char**
 }
 
 QueryNode* RediSearch_CreateTokenNode(IndexSpec* sp, const char* fieldName, const char* token) {
+  if (StopWordList_Contains(sp->stopwords, token, strlen(token))) {
+    return NULL;
+  }
```
Comment (ashtul) [inline src/redisearch_api.c:256]: This is what fixes RedisGraph/RedisGraph#1638 issue

File: tests/cpptests/test_cpp_llapi.cpp [hunk 4]
```diff
@@ -820,3 +820,35 @@ TEST_F(LLApiTest, testScorer) {
...
+  RSQNode *node = RediSearch_CreateTokenNode(index, "doesnt_matter", "is");
+  ASSERT_EQ((size_t)node, 0);
+  node = RediSearch_CreateTokenNode(index, "doesnt_matter", "Redis");
+  ASSERT_NE((size_t)node, 0);
+  RediSearch_QueryNodeFree(node);
+  RediSearch_DropIndex(index);
+
+  // Check custom stopword list
+  const char *words[] = {"Redis", "Labs"};
+  RSIdxOptions opts1 = { .stopwords = words, .stopwordsLen = 2 };
```
Comment (MeirShpilraien) [inline tests/cpptests/test_cpp_llapi.cpp:839]: An index option is an opaque object, needs to be created with `RediSearch_CreateIndexOptions`. I believe it's a good idea that the tests will be using the LLAPI the same way the user will.

Comment (ashtul) [reply tests/cpptests/test_cpp_llapi.cpp:839]: You are completely right! :+1:

File: src/redisearch_api.c [hunk 5]
```diff
@@ -510,6 +518,12 @@ void RediSearch_IndexOptionsSetGetValueCallback(RSIndexOptions* options, RSGetVa
   options->gvcbData = ctx;
 }
 
+void RediSearch_IndexOptionsSetStopwords(RSIndexOptions* opts, const char **stopwords, int stopwordsLen) {
+  opts->stopwords = stopwords;
```
Comment (MeirShpilraien) [inline src/redisearch_api.c:522]: I think it a good idea to perform a full copy of this array, we have no idea if the given argument is stack-allocated or heap-allocated. It's not something that happened a lot and I think it's OK to copy it.

Comment (MeirShpilraien) [reply src/redisearch_api.c:522]: And then you also need to free it on RediSearch_FreeIndexOptions

**Review by MeirShpilraien — COMMENTED** [review]
Small comment

File: src/redisearch_api.c (hunk 5 above)
Comment (ashtul) [reply src/redisearch_api.c:522]: @MeirShpilraien  We don't need these strings. We just use them to populate a trie. This is why we use `const`.

Comment (MeirShpilraien) [reply src/redisearch_api.c:522]: Who promises those strings will not be free by the time you create the index? Who promises the user will not create one index option and will use it again and again to create indexes?

Comment (ashtul) [reply src/redisearch_api.c]: Done

**Review by MeirShpilraien — APPROVED** [review]

File: src/redisearch_api.c (hunk 3 above)
Comment (sidenAlex) [reply src/redisearch_api.c:257]: Can you please show me to use this in the RG commands? I cannot find documentation for this and I have tried guessing and checking several items with no success.

Comment (ashtul) [reply src/redisearch_api.c:257]: @sidenAlex
You might find our unit tests helpful. ([link](https://github.com/RediSearch/RediSearch/blob/b19e16f58fb8cb867444755710d55f2cd6107898/tests/cpptests/test_cpp_llapi.cpp#L824))

**Pull request merged**